randomgen/src/distributions/distributions.c

```diff
@@ -1220,42 +1222,86 @@ static NPY_INLINE uint64_t bounded_uint64(brng_t *brng_state, uint64_t off,
   return off + val;
 }
 
-uint64_t random_bounded_uint64(brng_t *brng_state, uint64_t off, uint64_t rng,
+uint64_t random_bounded_uint64(brng_t *brng_state,
+                               uint64_t off, uint64_t rng,
                                uint64_t mask) {
   return bounded_uint64(brng_state, off, rng, mask);
 }
 
-static NPY_INLINE uint32_t bounded_uint32(brng_t *brng_state, uint32_t off,
-                                          uint32_t rng, uint32_t mask) {
+static NPY_INLINE uint32_t bounded_masked_uint32(brng_t *brng_state,
+                                                 uint32_t off, uint32_t rng,
+                                                 uint32_t mask) {
   /*
    * The buffer and buffer count are not used here but are included to allow
    * this function to be templated with the similar uint8 and uint16
    * functions
    */
-
   uint32_t val;
-  if (rng == 0)
-    return off;
 
   while ((val = (next_uint32(brng_state) & mask)) > rng)
     ;
   return off + val;
 }
 
-uint32_t random_buffered_bounded_uint32(brng_t *brng_state, uint32_t off,
-                                        uint32_t rng, uint32_t mask, int *bcnt,
-                                        uint32_t *buf) {
+static NPY_INLINE uint32_t bounded_lemire_uint32(brng_t *brng_state,
+                                                 uint32_t off, uint32_t rng) {
+  /*
+   * Uses Lemire's algorithm - https://arxiv.org/abs/1805.10941
+   *
+   * The buffer and buffer count are not used here but are included to allow
+   * this function to be templated with the similar uint8 and uint16
+   * functions
+   */
+  const uint32_t rng_excl = rng + 1;
+  uint64_t m;
+  uint32_t leftover;
+
+  if (rng == 0xFFFFFFFF) {
+    return next_uint32(brng_state);
+    /* ToDo: Move this code to caller to prevent this check on each call when generating arrays of numbers. */
+  }
+
+  /* Generate a scaled random number. */
+  m = ((uint64_t)next_uint32(brng_state)) * rng_excl;
+
+  /* Rejection sampling to remove any bias */
+  leftover = m & ((uint32_t)((1ULL << 32) - 1));
+
+  if (leftover < rng_excl) {
+      const uint32_t threshold = ((uint32_t)((1ULL << 32) - rng_excl)) % rng_excl;
+
+      while (leftover < threshold) {
+          m = ((uint64_t)next_uint32(brng_state)) * rng_excl;
```

Comment:
There should be a decent optimization available for platforms w/o __uint128 since the upper bits only need to be computed after it is accepted. Still adds quite a few more multiplications and adds, but it isn't strictly necessary to compute the entire 128 bit in each draw, only the lower 64 bits of the product.

Reply:
Nice. I've made a note of that. Thanks.